src/ota_metadata/legacy/csv_parser.py

```diff
@@ -0,0 +1,281 @@
+# Copyright 2022 TIER IV, INC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Parse the CSV based OTA image metafiles and import into database."""
+
+
+from __future__ import annotations
+
+import re
+import stat
+from pathlib import Path
+
+from ota_metadata.file_table import (
+    FileEntryAttrs,
+    FileTableDirectories,
+    FileTableDirORM,
+    FileTableNonRegularFiles,
+    FileTableNonRegularORM,
+    FileTableRegularFiles,
+    FileTableRegularORM,
+)
+from otaclient_common.typing import StrOrPath
+
+from .rs_table import ResourceTable, ResourceTableORM
+
+ENTRIES_PROCESS_BATCH_SIZE = 2048
+
+
+def de_escape(s: str) -> str:
+    """de-escape the previously escaped '\' character."""
+    return s.replace(r"'\''", r"'")
+
+
+#
+# ------ dirs.txt CSV parser implementation ------ #
+#
+
+DIR_CSV_PATTERN = re.compile(r"(?P<mode>\d+),(?P<uid>\d+),(?P<gid>\d+),(?P<path>.*)")
+
+
+def parse_dirs_csv_line(line: str) -> FileTableDirectories:
+    """Directory entry's CSV pattern.
+
+    Examples:
+    0755,1000,1000,'.'
+    0755,1000,1000,'my-dir'
+    0755,1000,1000,'my-dir/my-sub-dir'
+    0755,1000,1000,'single'\''quote'
+    """
+    _ma = DIR_CSV_PATTERN.match(line.strip())
+    assert _ma is not None, f"matching dirs failed for {line}"
+
+    # NOTE: the ota-metadata generator strips away the file type bits,
+    #       we need to put it back into the mode.
+    #       os.chmod will just ignore the file type bits.
+    mode = int(_ma.group("mode"), 8) | stat.S_IFDIR
+    uid = int(_ma.group("uid"))
+    gid = int(_ma.group("gid"))
+    path = de_escape(_ma.group("path")[1:-1])
+
+    return FileTableDirectories(
+        path=path,
+        entry_attrs=FileEntryAttrs(
+            uid=uid,
+            gid=gid,
+            mode=mode,
+        ),
+    )
+
+
+def parse_dirs_from_csv_file(
+    _fpath: StrOrPath, _orm: FileTableDirORM, *, cleanup: bool = True
+) -> int:
+    """Compatibility to the plaintext CSV dirs.txt."""
+    _batch, _batch_cnt = [], 0
+    try:
+        with open(_fpath, "r") as f:
+            _idx = 0
+            for _idx, line in enumerate(f, start=1):
+                _new = parse_dirs_csv_line(line)
```

Comment:
If this parser is only for csv, how about to use the default CSV library instead of regex? I imagine the reason to use regex is to re-use the existing parser for text, but feel that csv library for csv parser is more safe.
Another idea is to aggregate regext module as helper function between text and csv. It might be preference, but I don't mind to create/use such utility functions between Classes.
https://docs.python.org/3/library/csv.html